### PR TITLE
Aha/vfs layer2

### PR DESCRIPTION
**Pull request by andyHa**

### CODE REVIEW

File: src/main/java/sirius/biz/storage/util/WatchableOutputStream.java [hunk 1]
```diff
@@ -9,9 +9,16 @@
 package sirius.biz.storage.util;
 
 import sirius.kernel.async.Future;
+import sirius.kernel.commons.Files;
```
Comment (tbiScireum) [inline src/main/java/sirius/biz/storage/util/WatchableOutputStream.java]: unnecessary ?

File: src/main/java/sirius/biz/storage/layer3/MutableVirtualFile.java [hunk 2]
```diff
@@ -71,6 +72,20 @@ public MutableVirtualFile withCanCreateChildren(Predicate<VirtualFile> canCreate
         return this;
     }
 
+    /**
+     * Determines if file is generally considered readonly.
```
Comment (sabieber) [inline src/main/java/sirius/biz/storage/layer3/MutableVirtualFile.java]:
```suggestion
     * Determines if the file is generally considered readonly.
```

File: src/main/java/sirius/biz/storage/layer3/VirtualFile.java [hunk 3]
```diff
@@ -731,6 +813,177 @@ public OutputStream createOutputStream() {
...
+            return true;
+        }
+
+        return internalCanConsumeFile();
+    }
+
+    /**
+     * Tries to consume the given stream to update the contents of this file.
+     * <p>
+     * Note that if the source i a {@link File} {@link #consumeFile(File)} can be used which is likely to be more
```
Comment (sabieber) [inline src/main/java/sirius/biz/storage/layer3/VirtualFile.java]: i a -> is a

File: src/main/java/sirius/biz/storage/layer3/VirtualFile.java [hunk 4]
```diff
@@ -731,6 +813,177 @@ public OutputStream createOutputStream() {
...
+            }
+        }
+
+        return false;
+    }
+
+    /**
+     * Consumes the given stream to update the contents of this file or throws an exception if this fails.
+     * <p>
+     * Note that if the source i a {@link File} {@link #consumeFile(File)} can be used which is likely to be more
```
Comment (sabieber) [inline src/main/java/sirius/biz/storage/layer3/VirtualFile.java]: i a -> is a

File: src/main/java/sirius/biz/storage/layer2/BlobStorage.java [hunk 5]
```diff
@@ -23,14 +26,31 @@
...
      * <p>
      * A browsable space has a root directory per tenant and can be viewed and used like a regular file system
-     * via the {@link L3Uplink}.
+     * via the {@link L3Uplink uplink}.
      */
     public static final String CONFIG_KEY_LAYER2_BROWSABLE = "browsable";
 
-    protected DerivedSpaceInfo<Boolean> editable = new DerivedSpaceInfo<>(CONFIG_KEY_LAYER2_BROWSABLE,
+    /**
+     * Determines if the space is mounted as readonly.
```
Comment (sabieber) [inline src/main/java/sirius/biz/storage/layer2/BlobStorage.java]: I think this javadoc and the one on `CONFIG_KEY_LAYER2_BROWSABLE` is a bit misleading, as the constants dont determine anything on their own. The DerivedSpaceInfos below do that.

File: src/main/java/sirius/biz/storage/layer3/VirtualFile.java (hunk 3 above)
Comment (sabieber) [inline src/main/java/sirius/biz/storage/layer3/VirtualFile.java]: i a -> is a

**Pull request merged**